pkg/build/controller/build/build_controller.go

```diff
@@ -966,23 +966,54 @@ func (bc *BuildController) resolveImageStreamTag(ref *corev1.ObjectReference, li
 		}
 		return nil, fmt.Errorf("the referenced image stream %s/%s could not be found: %v", namespace, name, err)
 	}
-	if newRef, ok := imageutil.ResolveLatestTaggedImage(stream, tag); ok {
+	newRef := ""
+	ok = false
+	if len(stream.Status.DockerImageRepository) > 0 {
+		newRef, ok = imageutil.ResolveLatestTaggedImage(stream, tag)
+	} else {
+		streamCopy := stream.DeepCopy()
+		streamCopy.Status.DockerImageRepository = bc.internalRegistryHostname
+		newRef, ok = imageutil.ResolveLatestTaggedImage(streamCopy, tag)
+	}
+	if ok {
 		// generate informational event if
 		// a) tag has spec
 		// b) tag is local reference
-		// c) status DockerImageRepository is not set, meaning we got this copy of the stream from the watch
+		// c1) status DockerImageRepository is not set, meaning we got this copy of the stream from the watch
 		//    before the image registry was set up
+		// c2) or the import has not happened yet
 		// note, this method is only called for non-output image refs, so we should not get duplicates with
 		// the error/warning currently in place for the output link falling into this category
 		ref2, ok2 := imageutil.SpecHasTag(stream, tag)
-		if ok2 && ref2.ReferencePolicy.Type == imagev1.LocalTagReferencePolicy && len(stream.Status.DockerImageRepository) == 0 {
-			bc.recorder.Eventf(build, corev1.EventTypeNormal,
-				"LocalRefImageStreamTagUnavailable",
-				"The build %s/%s references an image stream tag %s with the local reference policy type, but the internal image registry location was not marked in the image stream",
-				build.Namespace,
-				build.Name,
-				ref.Name)
-		}
+		if ok2 && ref2.ReferencePolicy.Type == imagev1.LocalTagReferencePolicy {
+			if len(stream.Status.DockerImageRepository) == 0 {
+				bc.recorder.Eventf(build, corev1.EventTypeNormal,
+					"LocalRefImageStreamTagUnavailableNoImageRegistryRef",
+					"The build %s/%s references an image stream tag %s with the local reference policy type, but the internal image registry location was not marked in the image stream",
+					build.Namespace,
+					build.Name,
+					ref.Name)
+			}
+			latestTaggedImage := imageutil.LatestTaggedImage(stream, tag)
+			if len(latestTaggedImage.Image) == 0 {
```

Comment:
i'm not understanding the purpose of this...  if the import hasn't happened/succeeded yet, won't `imageutil.ResolveLatestTaggedImage(streamCopy, tag)` have already returned you nil/an error?

Reply:
https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L247

Reply:
I'm worried about this conditional at https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L333

Some of the surrounding context around that `if`:

```
	case imagev1.LocalTagReferencePolicy:
		local := stream.Status.DockerImageRepository
		if len(local) == 0 || len(latest.Image) == 0 {
			// fallback to the originating reference if no local docker registry defined or we
			// lack an image ID
			return latest.DockerImageReference
		}
```

Do we think the `|| len(latest.Image) == 0` is unnecessary there?

If not, then my additional alert test tells us if we had a situation where `len(stream.Status.DockerImageRepository)` indicated the img reg was set, but since `latest.Image` was not populated, we still did not set up local pullthrough.

Reply:
> Do we think the || len(latest.Image) == 0 is unnecessary there?

i think it's unnecessary in your specific flow.  It may be useful to other callers.  

you are calling this:
https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L293

which calls this:
https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L238

which checks this and returns nil if there are no valid entries associated w/ the tag:
https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L246-L247

which means when it then calls this:
https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L297

it will return this:
https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L304

which means you will never enter the if block here:
https://github.com/openshift/openshift-controller-manager/pull/34/files#diff-4349d2367514c13a1d5c76dc4debf58fR978

so you'll never get to:
https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L333

Whether that line is unnecessary in general i can't say, there might be other callers of https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L321 that come in with different arguments.

> If not, then my additional alert test tells us if we had a situation where len(stream.Status.DockerImageRepository) indicated the img reg was set, but since latest.Image was not populated, we still did not set up local pullthrough.

if latest.image is not populated, nothing else matters.  if latest.image is not populated, there is no pullspec/sha available, which is a scenario this code already handled and errors reported.  If you run a build for a buildconfig that points to an imagestream that has not successfully imported and you'll see it:

```
$ oc describe is bar
Name:			bar
Namespace:		p1
Created:		4 minutes ago
Labels:			<none>
Annotations:		openshift.io/image.dockerRepositoryCheck=2019-10-11T17:45:05Z
Image Repository:	image-registry.openshift-image-registry.svc:5000/p1/bar
Image Lookup:		local=false
Unique Images:		0
Tags:			2

latest
  tagged from docker.io/foo/bar

  ! error: Import failed (Unauthorized): you may not have access to the container image "docker.io/foo/bar:latest"
      4 minutes ago

test
  tagged from docker.io/foo/bar:test

  ! error: Import failed (Unauthorized): you may not have access to the container image "docker.io/foo/bar:test"
      18 seconds ago

$ oc start-build nodejs-ex
The ImageStreamTag "bar:test" is invalid: from: Error resolving ImageStreamTag bar:test in namespace p1: unable to find latest tagged image
```

Reply:
Thanks for all the detail, that helps.

Also, I *finally* found the decoration stuff you mentioned earlier.  I've looked at it before, but could not remember where it was exactly.  It is part of etcd storage.

So bottom line, the `stream.Status.DockerImageRepository` gets set on ImageStream creation, update, and it not tied to the IST imports in any way (which is what was the core of what I was worried about when thinking if it was possible for the setting of those items at https://github.com/openshift/library-go/blob/master/pkg/image/imageutil/helpers.go#L333 to somehow be related).

I'll clean up the unneeded checks and update the PR momentarily.